### PR TITLE
劣モジュラ関数によるエントリの部分集合選択のサポート

### PR DESCRIPTION
**Pull request by syou6162**
- [A class of submodular functions for document summarization](http://dl.acm.org/citation.cfm?id=2002537)
- [スライド](http://melodi.ee.washington.edu/~bilmes/mypubs/lin2011-class-submod-sum.slides.pdf)、[解説](http://www.lr.pi.titech.ac.jp/~morita/YANS.pdf)
- [#30 A class of Submodular Functions for Document Summarization - yasuhisa's blog](http://www.yasuhisay.info/entry/20110720/1310437007)

### CODE REVIEW

**Comment by syou6162** [comment]
劣モジュラなしの20件の結果。

```console
% make && ./go-active-learning apply --input-filename <(cat ml_train.tsv tweet.txt additional_ml_*.txt | sort | uniq) --filter-status-code-ok --json-output | jq -s -r 'sort_by(.Score) | reverse | .[] | select(.Score > 0) | "- " + (.Score | tostring) + ": [" + .Title + "](" + .Url + ")"' | head -n 20
```

- 196.13043478260894: [CycleGAN](https://t.co/I4dUlKNuUS)
- 163.70931677018635: [Bridging the Gap between   Human and Machine Translation](https://t.co/iQkH3Ge6V0)
- 143.807453416149: [[1705.08520] An effective algorithm for hyperparameter optimization of neural   networks](https://t.co/x5mJHMniGf)
- 137.41614906832285: [Everything that Works Works Because it's  Bayesian](https://t.co/S1xKMFY1Fs)
- 133.2360248447205: [an Easy-to-Use Named-Entity Recognition Tool based on Neural Networks](https://t.co/mb8OvseKf4)
- 131.94782608695644: [音声の声質を変換する技術とその応用](https://t.co/jKsMaagglV)
- 127.97639751552795: [[1705.07204] Ensemble Adversarial Training](https://t.co/Ufx0Ky4ZMK)
- 125.23726708074531: [[1705.08292v1] The Marginal Value of Adaptive Gradient Methods in Machine Learning](https://t.co/uVJ0KMen5N)
- 117.77515527950308: [Towards Deeper Understanding of Moment Matching Network](https://t.co/34ZAjACEum)
- 116.56645962732905: [Kullback-Leibler Divergence Explained](https://t.co/RboMuGBGqT)
- 115.16770186335404: [A Hybrid Temporal Convolutional and Recurrent Network for   Video Action Segmentation](https://t.co/RnZXnPJcJU)
- 114.41987577639749: [Invited Speakers](https://t.co/h9seCrdZDe)
- 113.42857142857152: [High-Level Explanation of Variational Inference](https://t.co/6jlJruLv5L)
- 110.32173913043474: [[1705.08933] Doubly Stochastic Variational Inference for Deep Gaussian Processes](https://t.co/VbT8gRWkpK)
- 106.91677018633536: [[1705.08153] Visualizing LSTM decisions](https://t.co/T9YXyb3yu7)
- 101.75031055900621: [[1705.08039] Poincar\'e Embeddings for Learning Hierarchical Representations](https://t.co/C2TYfz81bM)
- 94.24720496894408: [Generating Code from a Graphical User Interface Screenshot](https://t.co/HtX18r1LI5)
- 92.96770186335397: [[1705.07606] Deep Reinforcement Learning with Relative Entropy Stochastic Search](https://t.co/dDcGJNNtHi)
- 92.75155279503113: [An early overview of ICLR2017](https://t.co/bcKYiIkF1g)
- 86.01987577639748: [[1705.07177] Model-Based Planning in Discrete Action Spaces](https://t.co/8NHTJxyfdq)

**Comment by syou6162** [comment]
劣モジュラありの結果。rにsensitiveではある。

```console
make && ./go-active-learning apply --input-filename <(cat ml_train.tsv tweet.txt additional_ml_*.txt | sort | uniq) --filter-status-code-ok --json-output --subset-selection -r 0.75 --size-constraint 20 | jq -s -r 'sort_by(.Score) | reverse | .[] | select(.Score > 0) | "- " + (.Score | tostring) + ": [" + .Title + "](" + .Url + ")"'
```

- 127.87760097919217: [[1705.08520] An effective algorithm for hyperparameter optimization of neural   networks](https://t.co/x5mJHMniGf)
- 110.80905752753976: [[1705.08292v1] The Marginal Value of Adaptive Gradient Methods in Machine Learning](https://t.co/uVJ0KMen5N)
- 100.61566707466345: [Generating Code from a Graphical User Interface Screenshot](https://t.co/HtX18r1LI5)
- 100.32435740514076: [A Hybrid Temporal Convolutional and Recurrent Network for   Video Action Segmentation](https://t.co/RnZXnPJcJU)
- 96.91799265605871: [[1705.08933] Doubly Stochastic Variational Inference for Deep Gaussian Processes](https://t.co/VbT8gRWkpK)
- 42.28029375764992: [[1705.08918] Unsupervised Learning Layers for Video Analysis](https://t.co/HZITQj304f)
- 21.57772337821297: [fchollet/keras-resources: Directory of tutorials and open-source code repositories for working with Keras, the Python deep learning library](https://t.co/viFaxT1Ktu)
- 16.905752753977968: [Deep Furiganaを機械学習で自動でふる](https://t.co/93Lt1GxNbs)
- 16.315789473684205: [機械学習関連カンファレンスの論文締め切り直前にGoogleやMSのクラウドサービスのGPUが一時的に枯渇](https://t.co/9depszOt9u)
- 9.801713586291312: [機械学習の情報を手法を中心にざっくり整理](http://qiita.com/ishizakiiii/items/f6909696c616fd6294ca#%E6%89%8B%E6%B3%95)
- 4.376988984088127: [ニューラルネットワークの学習の工夫](http://s0sem0y.hatenablog.com/entry/2017/01/31/041148)
- 3.2080783353733153: [AIRS 2017](https://t.co/sUapIJmt03)
- 2.2166462668298625: [実世界の人工知能@DeNA TechCon 2017](http://www.slideshare.net/pfi/dena-techcon-2017)
- 2.1077111383108944: ["事後分布，というかMAP推定値の予測性能だけを見て，ハイパーパラメータやモデルの選択ができればよいのだけれど，典型的には漸近論が成り立つとはいいがたい状況なのがつらいところ（これもまた全然わかってない人がいて泣けるん ですが．いつかちゃんと解説書かないといけないか） https](https://t.co/CEeKaAzaKb)
- 1.8029375764993882: [「深層学習による自然言語処理」を書きました](https://t.co/CVbQBcyZNV)
- 1.3757649938800505: [深層学習：ハイパーパラメータの設定に迷っている人へ](http://s0sem0y.hatenablog.com/entry/2016/11/13/035443)
- 0.4320685434516516: ["図書館での「ぬいぐるみお泊まり会」の効果を検証する論文が、なんかすげくかわいい。  > ぬいぐるみお泊まり会：読書活動の促進と効果の持続性 https](https://t.co/cdykioKo7k)
- 0.34883720930232576: [結局、機械学習に必要な数学ってなに？](http://futurismo.biz/archives/6106)
- 0.08445532435740455: [Deep Learning with Keras](https://t.co/cd5xZnndAO)
- 0.07588739290085722: ["@ibaibabaibai ほんこれは「本当にこれだわ」って意味で賛同を意味してます。  あと、最後のベイズ体系が、よくわからん生成モデルとコツコツ統計モデルの中で微妙な立場というのがよくわからないのですが、それは今までのベイズ の使われ方の意味で言ってますか？"](https://t.co/EOl7Pw2D38)

**Comment by syou6162** [comment]
論文では目的関数に多様性を担保する部分が組込まれているが、そのために高次元でもそこそこちゃんと動くk-meansをやるのもダルいし、実験結果的にも多様性なしとスコアはほぼ変わっていないので、今回はなしでいいのではないか。
